dlib/math/windows.h

```diff
@@ -0,0 +1,272 @@
+    };
+
+    struct window_length
+    /*!
+        This ojbect is a strong type representing the window length of a kaiser window.
+        See https://en.wikipedia.org/wiki/Kaiser_window.
+    !*/
+    {
+        explicit window_length(std::size_t N_) : N{N_} {}
+        std::size_t N = 0;
+    };
+
+    struct symmetric_t {};
+    /*!
+        This object is a strong type that signifies that the window is a symmetric window.
+        See https://en.wikipedia.org/wiki/Window_function#Symmetry for a discussion on
+        symmetric vs periodic windows
+    !*/
+
+    struct periodic_t {};
+    /*!
+        This object is a strong type that signifies that the window is a periodic window.
+        See https://en.wikipedia.org/wiki/Window_function#Symmetry for a discussion on
+        symmetric vs periodic windows
+    !*/
+
+    // ----------------------------------------------------------------------------------------
+
+    template<typename R>
+    inline R kaiser(R x, window_duration L, beta_t beta)
+    /*!
+        This computes the kaiser window function or kaiser-bessel window function.
+        See https://en.wikipedia.org/wiki/Kaiser_window
+
+        ensures
+            - returns the kaiser window function when |x| <= L/2 where L is the window duration
+            - returns 0 otherwise
+    !*/
+    {
+        static_assert(std::is_floating_point<R>::value, "template parameter must be a floating point type");
+
+        if (std::abs(x) <= L.L/R{2})
+        {
+            const R r = 2*x/L.L;
+            const R a = dlib::cyl_bessel_i(0, beta.v*sqrt(1-r*r));
+            const R b = dlib::cyl_bessel_i(0, beta.v);
+            return a / b;
+        }
+        else
+        {
+            return R{0};
+        }
+    }
+
+    template<typename R>
+    inline R kaiser(index_t i, window_length N, beta_t beta, symmetric_t)
+    /*!
+        This computes the kaiser window function or kaiser-bessel window function.
+        See https://en.wikipedia.org/wiki/Kaiser_window
+        This variant is a short-cut for computing a window function and storing it
+        in an array of size N where 0 <= i < N is the array index.
+        This is the symmetric version.
+
+        ensures
+            - returns kaiser_r(i - (N-1)/2, N-1, beta)
+    !*/
+    {
+        static_assert(std::is_floating_point<R>::value, "template parameter must be a floating point type");
+        DLIB_ASSERT(i.i < N.N, "index out of range");
+        return kaiser(R(i.i) - R(N.N-1) / R(2), window_duration{R(N.N-1)}, beta);
+    }
+
+    template<typename R>
+    inline R kaiser(index_t i, window_length N, beta_t beta, periodic_t)
+    /*!
+        This computes the kaiser window function or kaiser-bessel window function.
+        See https://en.wikipedia.org/wiki/Kaiser_window
+        This variant is a short-cut for computing a window function and storing it
+        in an array of size N where 0 <= i < N is the array index.
+        This is the periodic version.
```

Comment:
The link https://en.wikipedia.org/wiki/Window_function#Symmetry explains it quite well I think. It follows the same nomenclature as Matlab, Scipy and probably most DSP libraries.

Reply:
I'll need to add at some point overloads of these functions for fillings arrays, vectors and matrices. But I don't think those overloads belong in the `math` directory. Otherwise you're polluting the math headers with STL container headers and the dlib matrix stuff which pulls in quite a lot of stuff.